### ISSUE TITLE
Question on kernel_size in decoder

### ISSUE THREAD

**Issue by iliaschalkidis** [liangstein/ByteNet-Keras#1]
Hi @liangstein,

Thanx a lot for your implementation, it helped me a lot to understand bytenet.

I have a technical question for your implementation. I see that you use kernel_size=3 in your encoder's Convolutions, which makes absolute sense with padding='same' in order to condition each timestep (t) based on timesteps (t-1, t+1)*. My question is why the kernel_size=3 in decoder's Convolutions, where you use padding='causal' and you only need to condition based on the previous timestep (t-1). Is this the correct size or you should probably use kernel_size=2 in order to avoid including extra timesteps (e.g., timestep t-2)?

Thanx, Ilias

* where timestep is the previous word in the first residual block, while it is the context-aware embedding of your current word in the next blocks.

**Comment by liangstein** [comment]
I don't think conv filter size is a problem, the vital thing is the alignment of the output sequence. @iliaschalkidis

**Comment by iliaschalkidis** [comment]
@liangstein  In file keras/backend/tensorflow_backend.py line 3513-3515, it seems that when you use kernel_size=3 with padding="causal",  the tensorflow backend implementation creates a left_pad mask by turning kernel_size to 2 (kernel_shape[0] - 1) and padding to "valid". It actually internally interprets your initial choices (kernel_size=3, padding="causaul") into (kernel_size=2, padding="valid") in order to create the preferable causal dilation effect as it is being described in the article.

**Comment by liangstein** [comment]
@iliaschalkidis I didn't dig into this far. However from my experience with Keras, "causal" means the output at time step t is only related to outputs from 0 to (t-1). Did you mean if I choose kernel_size=3 with "causual", as the padding is transfered to "valid" by tensorflow, the output at t is not related to the outputs from 0 to (t-1) ? or the output is no longer a sequence output?

**Comment by iliaschalkidis** [comment]
@liangstein I think what really happens is that while you define kernel_size=3 (trigrams), the backend actually works with bigrams on timesteps (t, t-1), which is what you actually really wanted to do in the first place (I suspect there is such thing as causal padding in tensorflow, so they have to interpret you code somehow internally in order to make the same effect  with valid). 

Trigrams makes sense on the encoder when you condition on both past and future tokens, so you actually start with token (t-1,t,t+1). I was afraid that keras would also interpret your kernel size of decoder into filters of trigrams (t-2, t-1, t) with causal padding, which is not terrible but it is not what bytenet does. The code actually proves that it works as  both of us excepted in terms of article's replicability: trigrams on encoder, bigrams on decoder. 

I'm testing your decoder as an encoder for a CNN-based language model that may turn to be 10x faster than training LSTMs with minimal  difference in performance, which is super important for people that do not have a server farm :)

**Comment by liangstein** [comment]
@iliaschalkidis  I think Keras must have transfered to (t-2,t-1,t) somewhere, and it also works.
